pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -139,112 +153,293 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
           waitForStable(tableNameWithType);
           // clear retries as it tracks failures with each idealstate update attempt
           retries = 0;
+          LOGGER.info("External view converged for the change in ideal state for table {}. Will start the next iteration (if any)",
+              tableNameWithType);
           continue;
         }
-        // in case of any error, we retry a bounded number of types
+        // in case of any error, we retry a bounded number of times
       } catch (ZkBadVersionException e) {
-        LOGGER.warn("Version changed while updating ideal state for resource: {}", tableNameWithType);
+        // we will go back in the loop and reattempt by recomputing the target ideal state
+        LOGGER.info("Version changed while updating ideal state for resource: {}, was expecting version {}",
+            tableNameWithType, currentIdealState.getRecord().getVersion());
       } catch (Exception e) {
-        LOGGER.warn("Caught exception while updating ideal state for resource: {}", tableNameWithType, e);
+        if (e instanceof IllegalStateException && e.getCause() instanceof ExternalViewErrored) {
+          if (e.getCause() instanceof ExternalViewErrored) {
+            LOGGER.error("External view reported error for table {} after updating ideal state", tableNameWithType);
+          } else if (e.getCause() instanceof ExternalViewConvergeTimeout) {
+            LOGGER.error("Timedout while waiting for external view to converge for table {}", tableNameWithType);
+          }
+          // remove the cause as it is private and is only used to detect exact error within this class
+          throw new IllegalStateException(e.getMessage());
+        } else {
+          LOGGER.error("Caught exception while or after updating ideal state for resource: {}", tableNameWithType, e);
+          throw e;
+        }
       }
 
+      // if we are here, we failed to persist the updated ideal state in ZK
+      // due to version mismatch, will attempt again
       previousIdealState = currentIdealState;
       if (retries++ > MAX_RETRIES) {
         LOGGER.error("Unable to rebalance table {} in {} attempts. Giving up", tableNameWithType, MAX_RETRIES);
+        result.setStatus("Cancelling the rebalance operation as we have exhausted the max number of retries after ZK version mismatch");
+        result.setStatusCode(RebalanceResult.RebalanceStatus.FAILED);
         return result;
       }
       // wait before retrying
       try {
         Thread.sleep(retryDelayMs);
       } catch (InterruptedException e) {
         LOGGER.error("Got interrupted while rebalancing table {}", tableNameWithType);
+        result.setStatus("Rebalancer got interrupted");
+        result.setStatusCode(RebalanceResult.RebalanceStatus.FAILED);
         Thread.currentThread().interrupt();
         return result;
       }
     }
   }
 
+  private void validateMinReplicas(final String tableNameWithType,
```

Comment:
I think that checking replication factor from table config should be enough for validation.

Reply:
Done. No longer iterating over ideal state. Just using idealstate.getReplicas()

Reply:
I was mentioning `replication` and `numReplicasPerPartition` from table config. If you want to use `idealstate.getReplicas()`, I think that you should double check on how it gets updated. I think that we usually don't depend on that number.

Reply:
Initial thought was to use one of the methods from table config but @mcvsubbu suggested to use the ideal state method. I can see that setReplicas() method on ideal state is being invoked through PinotTableRestletResource -> PinotHelixResourceManager -> PinotTableIdealStateBuilder -> IdealStateBuilder. In my unit tests I see this happening every time a post request is sent to the controller for creating the table. In fact the value that is being passed down to IdealState.setReplicas() is coming from table config's SegmentValidationAndRetetionConfig.getReplication()

Reply:
Checking from table config is a bit involved, different for offline, hlc and llc. This check is done during table creation, and the idealstate is created with the right number of replicas. Since we are getting idealstate here anyway, it makes sense to check for replicas on the idealstate